### PR TITLE
 Email service, SMS service integrated with config service. 

### PR DESCRIPTION
**Pull request by Sparsh-Kumar**
## Description
- Completed Email service of communication service and it's test cases.
- Added test case to ensure that If provide invalid email (recipient, sender) then email service will throw error.
- Added test case to ensure that if provide empty sender name , then email service will throw error.
- Added test case to ensure that if all details provided(recipient email, sender email, sender name) are valid, then email service will send an email.
- Completed Sms service of communication service and it's test cases.
- Added test case to ensure to not allow sending sms, if recipient phone number is invalid.
- Added test case to ensure to not allow sending sms, if sms body is empty.

## Database schema changes
- N/A

### Automated test cases added
- should not allow to send email, if recipient email is invalid.
- should not allow to send email, if sender email is invalid.
- should not allow to send email, if sender name is empty.
- should allow to send email,  if recipient email, sender email, sender name are valid.
- should not allow to send sms, if the reciepient phone number is invalid
- should not allow to send sms, if the message body is empty.

### Manual test cases run
- N/A

### CODE REVIEW

File: src/modules/communication/email-service.ts [hunk 1]
```diff
@@ -0,0 +1,10 @@
+import SendGridService from './internals/send-grid-service';
+import { SendEmailParams } from './types';
+import { SendGridEmailParams } from './internals/sendgrid-email-params';
+export default class EmailService {
+  public static async sendEmail(params: SendEmailParams): Promise<void> {
+    const sendGridParams = new SendGridEmailParams(params);
```
Comment (jjalan) [inline src/modules/communication/email-service.ts:6]: Validation should happen inside `SendGridService.sendEmail`. The root level service should be implementation free

File: src/modules/communication/index.ts [hunk 2]
```diff
@@ -0,0 +1,11 @@
+import EmailService from './email-service';
+import SMSService from './sms-service';
+import { SendGridValidationError } from './types';
+import SendGridService from './internals/send-grid-service';
+
+export {
+  EmailService,
+  SMSService,
+  SendGridValidationError,
```
Comment (jjalan) [inline src/modules/communication/index.ts:9]: Nothing about sendgrid should be visible to outside world from here

File: src/modules/communication/internals/send-grid-service.ts [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+
+export default class SendGridService {
+  public static initializeService() {
+    const sendgridAPIKey:string = config.get('sendgridAPIKey'); // Todo: here use configuration service to get these details.
+    if(sendgridAPIKey) {
+      mail.setApiKey(sendgridAPIKey);
+    }
+  }
+  public static async sendEmail(params: SendEmailParams): Promise<void> {
+    const msg: any = {
```
Comment (jjalan) [inline src/modules/communication/internals/send-grid-service.ts:13]: Do not use `any`

File: src/modules/communication/internals/send-grid-service.ts [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+    msg.templateId = params.templateId;
+    msg.dynamic_template_data = params.templateData;
+    msg.dynamic_template_data = Object.assign(msg.dynamic_template_data, {
+      domain: config.get('webAppHost'),
+    });
+    try {
+      await mail.send(msg);
+    } catch(error) {
+      // Todo: Use logger Service to log these values
+      console.log(error.message);
```
Comment (jjalan) [inline src/modules/communication/internals/send-grid-service.ts:29]: We should throw error so the caller knows something went wrong and can take corrective action.

We should create ThirdPartyServiceError and throw its instance from here

**Review by jjalan — CHANGES_REQUESTED** [review]

File: src/modules/communication/internals/sendgrid-email-params.ts [hunk 5]
```diff
@@ -0,0 +1,59 @@
+import {
+  LooseObject,
+  SendEmailParams,
+  SendGridValidationError,
+  ValidationFailure
+} from "../types";
+
+export class SendGridEmailParams {
```
Comment (jjalan) [inline src/modules/communication/internals/sendgrid-email-params.ts:8]: This is not specific to send grid I believe, best to rename it to SendEmailPArams

File: src/modules/communication/internals/sendgrid-email-params.ts [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+        })
+      }
+      if(!isfromNameValid) {
+        failures.push({
+          field: 'fromName',
+          message: 'Please specify a valid name of the sending recipient.',
+        }); 
+      }
+      throw new SendGridValidationError(
+        'Cannot send email, as one or more params are invalid.',
```
Comment (jjalan) [inline src/modules/communication/internals/sendgrid-email-params.ts:48]: I have requested in the past on how to write good error message. Please Google about it

File: test/specs/sendgrid.spec.ts [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+  });
+  it('should send email, if all the values are valid.', async function() {
+    const params = {
+      to: 'to@gmail.com',
+      from: 'from@gmail.com',
+      fromName: 'Sample Name',
+      templateId: 'd-bea6fc9b30d34caeb73c11e2770e3287',
+      templateData: {}
+    };
+    sinonSandbox.stub(SendGridService, <any>'sendEmail').returns(Promise.resolve('1'));
```
Comment (jjalan) [inline test/specs/sendgrid.spec.ts:90]: Poor code spacing. Need line breaks

File: src/modules/communication/internals/send-grid-service.ts [hunk 8]
```diff
@@ -0,0 +1,32 @@
+import { SendEmailParams } from '../types';
+import config from 'config';
+import mail from '@sendgrid/mail';
+
+export default class SendGridService {
+  public static initializeService() {
+    const sendgridAPIKey:string = config.get('sendgridAPIKey'); // Todo: here use configuration service to get these details.
+    if(sendgridAPIKey) {
```
Comment (jjalan) [inline src/modules/communication/internals/send-grid-service.ts:8]: This does not sound right, if the caller has asked SendGridService.initializeService and if config is not, it should throw.

Failing silently results in unexpected behaviours.

File: src/modules/communication/internals/sendgrid-email-params.ts [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+    this.templateId = params.templateId;
+    this.templateData = params.templateData;
+  }
+
+  async validate() {
+    const failures: ValidationFailure[] = [];
+    const isToValid = this.isEmailValid(this.to);
+    const isFromValid = this.isEmailValid(this.from);
+    const isfromNameValid = !!this.fromName;
+    if(!isToValid || !isFromValid || !isfromNameValid) {
```
Comment (jjalan) [inline src/modules/communication/internals/sendgrid-email-params.ts:28]: Why is this needed?

Can we simply not check each ifs and push errors.

File: src/modules/communication/internals/sendgrid-email-params.ts [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+      }
+      throw new SendGridValidationError(
+        'Cannot send email, as one or more params are invalid.',
+        failures
+      );
+    }
+  }
+
+  isEmailValid(email: string) {
+    const re =
```
Comment (jjalan) [inline src/modules/communication/internals/sendgrid-email-params.ts:55]: Create a constants file and store the regex there.

File: src/modules/communication/index.ts (hunk 2 above)
Comment (Sparsh-Kumar) [reply src/modules/communication/index.ts:9]: @jjalan  sir, SendGridValidationError, is just another class initiated by me, it is not from sendgrid. This class is defined in types.ts.

Comment (jjalan) [reply src/modules/communication/index.ts:9]: Understood, it cannot be in index.ts file for the module.

Sendgrid is an internal detail

File: src/modules/communication/internals/sendgrid-email-params.ts [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+          message: 'Please specify a valid from email.'
+        })
+      }
+      if(!isfromNameValid) {
+        failures.push({
+          field: 'fromName',
+          message: 'Please specify a valid name of the sending recipient.',
+        }); 
+      }
+      throw new SendGridValidationError(
```
[the remark this answers is not in the document]
Comment (Sparsh-Kumar) [reply src/modules/communication/internals/sendgrid-email-params.ts:47]: the ```sendGridValidationError``` is just a class defined by me, It is not related to sendgrid , we are not throwing anything related to sendgrid to client.

[the remark this answers is not in the document]
Comment (jjalan) [reply src/modules/communication/internals/sendgrid-email-params.ts:47]: Who will catch this error?

[the remark this answers is not in the document]
Comment (Sparsh-Kumar) [reply src/modules/communication/internals/sendgrid-email-params.ts:47]: @jjalan  sir, actually I think the sendEmail service would be used by other service, so any controller which would make instance of this sendgrid-email-params, would contain a try catch to catch this error.

[the remark this answers is not in the document]
Comment (Sparsh-Kumar) [reply src/modules/communication/internals/sendgrid-email-params.ts:47]: ok, @jjalan  sir, got it, I will fix that, please also let me know if you have any workaround for this.

File: src/modules/communication/constant.ts [hunk 12]
```diff
@@ -0,0 +1 @@
+export const emailRegex = /^(([^<>()[\]\\.,;:\s@"]+(\.[^<>()[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;
```
Comment (jjalan) [inline src/modules/communication/constant.ts:1]: This is an internal detail, should be inside internal folder

File: src/modules/communication/index.ts [hunk 13]
```diff
@@ -0,0 +1,9 @@
+import EmailService from './email-service';
+import SMSService from './sms-service';
+import SendGridService from './internals/send-grid-service';
+
+export {
+  EmailService,
+  SMSService,
+  SendGridService,
```
Comment (jjalan) [inline src/modules/communication/index.ts:8]: SendGridService should not be exported

File: src/modules/communication/types.ts [hunk 14]
```diff
@@ -0,0 +1,57 @@
...
+  to: string;
+};
+
+export type SendSMSParams = {
+  to: PhoneNumber,
+  message: string
+};
+
+export enum ErrorCodes {
+  VALIDATION_ERROR = 404,
```
Comment (jjalan) [inline src/modules/communication/types.ts:24]: Please follow other module convention for error codes

File: src/modules/communication/types.ts [hunk 15]
```diff
@@ -0,0 +1,57 @@
...
+  [key: string]: any;
+};
+
+export type PhoneNumber = {
+  countryCode: string;
+  phoneNumber: string
+};
+
+export type SendEmailParams = {
+  from: string;
```
Comment (jjalan) [inline src/modules/communication/types.ts:11]: Best to combine `from`, `fromName` into 
```
export type EmailSender {
  email: string;
  name: string;
}
```
Same thing for to
```
export type EmailRecipient {
  email: string;
}
```

File: src/modules/communication/types.ts [hunk 16]
```diff
@@ -0,0 +1,57 @@
...
+
+export class MissingConfigVariable extends Error {
+  constructor(variableName: string) {
+    super(`Missing config variable ${variableName}.`);
+  }
+};
+
+export class ThirdPartyServiceError extends Error {
+  code: ErrorCodes;
+  constructor() {
```
Comment (jjalan) [inline src/modules/communication/types.ts:53]: you should pass third party name and error message from it in constructor. The error should be something like:

`${thirdParty} encountered error: ${thirdPartyErrorMessage}`

File: test/spec/email/sendgrid.spec.ts [hunk 17]
```diff
@@ -0,0 +1,106 @@
...
+  EmailService,
+} from '../../../src/modules/communication';
+import { initApp } from '../../helpers/spec-helpers';
+import mail from '@sendgrid/mail';
+
+chai.use(chaiHttp);
+let expressApp: Application;
+let sinonSandbox: sinon.SinonSandbox;
+
+describe('Email Service.', function() {
```
Comment (jjalan) [inline test/spec/email/sendgrid.spec.ts:16]: Specs are hard to read, there are no logical line breaks.

Ex: See this - https://github.com/sendgrid/sendgrid-nodejs/blob/78dae9ed05cc54bd50b4c0354a5e15855d83db60/packages/helpers/classes/response-error.js

You will see how there is logical chunk of separation.
